src/epgsql.erl

```diff
@@ -298,21 +298,38 @@ sync(C) ->
 cancel(C) ->
     epgsql_sock:cancel(C).
 
-%% misc helper functions
+%% @doc Execute a function within a transaction.
 -spec with_transaction(connection(), fun((connection()) -> Reply)) ->
-                              Reply | {rollback, any()}
-                                  when
-      Reply :: any().
+                          Reply | {rollback, Reply}
+                            when Reply :: any().
 with_transaction(C, F) ->
-    try {ok, [], []} = squery(C, "BEGIN"),
-        R = F(C),
-        {ok, [], []} = squery(C, "COMMIT"),
-        R
-    catch
-        _:Why ->
-            squery(C, "ROLLBACK"),
-            %% TODO hides error stacktrace
-            {rollback, Why}
+    {ok, [], []} = squery(C, "BEGIN"),
+    %% Result is always the result returned by the
+    %% function F. The code within the catch
+    %% will never return since it raises an exception.
+    Result = 
+      try
+        F(C)
+      catch
+        throw:Exception ->
+          {ok, [], []} = squery(C, "ROLLBACK"),
+          erlang:raise(throw, Exception, erlang:get_stacktrace());
+        error:Exception ->
+          {ok, [], []} = squery(C, "ROLLBACK"),
+          erlang:raise(error, Exception, erlang:get_stacktrace());
+        exit:Exception ->
+          {ok, [], []} = squery(C, "ROLLBACK"),
+          erlang:raise(exit, Exception, erlang:get_stacktrace())
+      end,
+    %% Only way to handle a rollback when executing commit.
+    Ref = epgsqli:squery(C, "COMMIT"),
+    CompleteCmd = receive {C, Ref, {complete, CompleteCommand}} -> CompleteCommand end,
+    receive {C, Ref, done} -> ok end,
```

Comment:
The problem here (just theorethical, but who knows?), is that epgsqli may send a lot of different messages (see https://github.com/epgsql/epgsql#simple-query
```
  {C, Ref, {columns, Columns}} ->
      %% columns description
      Columns;
  {C, Ref, {data, Row}} ->
      %% single data row
      Row;
  {C, Ref, {error, _E} = Error} ->
      Error;
  {C, Ref, {complete, {_Type, Count}}} ->
      %% execution of one insert/update/delete has finished
      {ok, Count}; % affected rows count
  {C, Ref, {complete, _Type}} ->
      %% execution of one select has finished
      ok;
  {C, Ref, done} ->
      %% execution of all queries from Sql has been finished
      done;
```
And we should be ready to handle any of them (how about `error` if worker was interrupted in time long-running heavy `COMMIT`).
That's not so trivial, as you can see here https://github.com/epgsql/epgsql/blob/3.2.0/test/epgsql_incremental.erl#L151